File: backend/src/main/java/com/resourcex/resourcex/service/PenaltyService.java
package com.resourcex.resourcex.service;

import com.resourcex.resourcex.dto.request.PenaltyRequest;
import com.resourcex.resourcex.dto.response.PenaltyResponse;
import com.resourcex.resourcex.entity.Penalty;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface PenaltyService {

    PenaltyResponse createPenalty(PenaltyRequest request);

    PenaltyResponse updatePenalty(Long penaltyId, PenaltyRequest request);

    PenaltyResponse getPenaltyById(Long penaltyId);

    Page<PenaltyResponse> getAllPenalties(Pageable pageable);

    List<PenaltyResponse> getPenaltiesByUserId(Long userId);

    List<PenaltyResponse> getPenaltiesByBookingId(Long bookingId);

    List<PenaltyResponse> getPenaltiesByDisputeId(Long disputeId);

    List<PenaltyResponse> getPenaltiesByStatus(Penalty.PenaltyStatus status);

    PenaltyResponse applyPenalty(Long penaltyId);

    PenaltyResponse waivePenalty(Long penaltyId);

    void deletePenalty(Long penaltyId);
}

File: frontend/web/lib/services/penaltyService.ts
import api from "../api";
import { PenaltyResponse } from "../../types/penalty";

export const penaltyService = {
  getAllPenalties: async (): Promise<PenaltyResponse[]> => {
    const response = await api.get<PenaltyResponse[]>("/penalties");
    return response.data;
  },

  getPenaltiesByUserId: async (userId: number): Promise<PenaltyResponse[]> => {
    const response = await api.get<PenaltyResponse[]>(`/penalties/user/${userId}`);
    return response.data;
  },

  createPenalty: async (penaltyData: {
    userId: number;
    bookingId?: number;
    disputeId?: number;
    amount: number;
    reason: string;
    issuedByUserId: number;
  }): Promise<PenaltyResponse> => {
    const response = await api.post<PenaltyResponse>("/penalties", penaltyData);
    return response.data;
  },

  applyPenalty: async (penaltyId: number): Promise<PenaltyResponse> => {
    const response = await api.patch<PenaltyResponse>(`/penalties/${penaltyId}/apply`);
    return response.data;
  },

  waivePenalty: async (penaltyId: number): Promise<PenaltyResponse> => {
    const response = await api.patch<PenaltyResponse>(`/penalties/${penaltyId}/waive`);
    return response.data;
  },
};
